Brave Today

kylehickinson opened the pull request:
Merge via merge-commit to keep history

## Summary of Changes

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes #2863

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue is assigned to a milestone (should happen at merge time).

kylehickinson committed:
Add Brave Today header

kylehickinson committed:
Add PrivateCDN image coder for SDWebImage

kylehickinson committed:
Fixes headline pair UI and add more cards from data source

kylehickinson committed:
Add context menu handling to feed items

iccub committed:
database initial setup

kylehickinson committed:
Fix PrivateCDN tests & decoder; Use padded images from feed

kylehickinson committed:
Add rest of iOS 13 actions & setup actions for all card types

Also adds correct sponsor card type and cleans up/documents some stuff

kylehickinson committed:
Update feed URL

iccub committed:
Add `urlHash` field, add remaining db methods.

iccub committed:
Basic insert test, add local db properties.

kylehickinson committed:
Add SDWebImage to DataTests "Copy Files" build phase

kylehickinson committed:
WIP: More work

kylehickinson committed:
Update feed source and add a CoreData model for it

kylehickinson committed:
Fixup new source MO

kylehickinson committed:
Disable sources since they arent working yet; Fix build error

iccub committed:
BraveTodayFeedItemMO simple unit test.

kylehickinson committed:
Generate cards for entire dataset and fix numbered group layout

Also:
- Fixes PrivateCDN crashing when downloading brand images
- Adds transition to image loads
- Fixes the brand image view not clipping (not totally fixed yet)

kylehickinson committed:
Setup proper layout for Brave Today on NTP and add target offset deceleration

kylehickinson committed:
Add branding text to groups when sources are not the same

kylehickinson committed:
Add and layout the Brave Today header

kylehickinson committed:
Fix up header button color and inset when displaying the header

kylehickinson committed:
Hide image if one doesn't exist for the item in the feed

- Removes unused `domain` in feed item content/MO

kylehickinson committed:
Only place feed items with images in headline/headline pair cards

Also:
- Perform card generation on background thread
- Fix potential issue where we may have no available articles for the a repeated sequence element. We will now exit early if no cards were generated from a repeated sequence

kylehickinson committed:
Handle opening when tapping (and via context menu)

Swap sponsors & deals filter

kylehickinson committed:
Calculate gap once instead of for each attribute

Also fix bug where Brave Today header disappears

kylehickinson committed:
Don't reload cards from new tabs if they've already loaded

kylehickinson committed:
Fix issue where items may not appear when scrolling up

kylehickinson committed:
Fix crash when attempting to create a headline pair card

kylehickinson committed:
Add better estimated sizes for automatic sized cells

kylehickinson committed:
Fix CV jump when you come back from tab tray

Also sets the scroll indicator style to white as its not visible in most circumstances

kylehickinson committed:
Fix-up action handlers for feed items to use UIAction extensions

kylehickinson committed:
Add slight color overlay on card background blurs

Replace Today welcome view icon

kylehickinson committed:
Give Today header bg blur the same treatment as cards

Update context menu bg on group items to be a bit brighter

kylehickinson committed:
Only start fading in the header halfway into the CV scroll

iccub committed:
Batch insert tests.

kylehickinson committed:
Add item hiding action handling

kylehickinson committed:
Move logic for menu building up to controller/section provider

Adds better support/a better entry point for legacy menus

kylehickinson committed:
Add iOS 12 legacy menu support

kylehickinson committed:
Cleanup Brave Today section provider

Use shared actions for all menu items, just switching containers

kylehickinson committed:
Dont show hide/block for non-article items

kylehickinson committed:
Add custom Brave symbols for hide item/block source menu actions

kylehickinson committed:
Adds source list UI

- Update source struct to match new server response
- Adds new tab and new private tab icons to iOS 13 context menu
- Updates DB classes as publisher name has moved from item to source

kylehickinson committed:
Group sources by Top News category then alpha. Adds section indexes

kylehickinson committed:
Fix Brave Today layout on iPads/landscape

kylehickinson committed:
Use brand text as fallback for logo in most brand placements

kylehickinson committed:
Remove hide and add disable source functionality

kylehickinson committed:
Remove hidden overlay for blocked sources

- Adds ability to re-enable sources from feed if they've been disabled

kylehickinson committed:
Fix margin spacing around custom SF Symbols

kylehickinson committed:
WIP: Brave Today

kylehickinson committed:
Display small modal alert when you disable a source

kylehickinson committed:
Replace first card graphic with lottie animation

kylehickinson committed:
Create error card and handle errors properly from Brave Today services

kylehickinson committed:
Support sources. Refactor FeedItem into Content & Source

kylehickinson committed:
Move back to fast deceleration rate for NTP

kylehickinson committed:
Add Brave Today settings

kylehickinson committed:
Update onboarding card copy

kylehickinson committed:
Add loader and small card animation when it completes loading

Add `skip_upload` option to enterprise fastlane build

kylehickinson committed:
Remove vertical scroll indicator on NTP

kylehickinson committed:
Add extra padding to end of Brave Today feed

This ensures that there's always enough space in the collection view to scroll the header into visibility

Also stops the intro card from showing while loading

kylehickinson committed:
Fix crash when choosing sources with no data loaded

kylehickinson committed:
Add specific card for deals and remove date from the items display

Also fix loader from showing up on opening new tab

kylehickinson committed:
Add categories to sources section in Brave Today settings

kylehickinson committed:
Update source DB model and remove feed item DB. Add source persistence

kylehickinson committed:
Add feed/source cache control and update logic

kylehickinson committed:
Add language check to determine Brave Today's initial default

kylehickinson committed:
Add Brave Today to Clear Private Data

kylehickinson committed:
Increase title label line cap to 5 for headlines/small headlines cards

kylehickinson committed:
Remove brand logo download + Cleanup content parsing

kylehickinson committed:
Increase line cap to 4 for both numbered and grouped vertical feed cards

kylehickinson committed:
Show disable source option for deals

kylehickinson committed:
Don't hide images for deals even if an image doesn't exist

kylehickinson committed:
Add fill strategies support for element sequences

kylehickinson committed:
Darken the Brave Today header bar by a small amount

kylehickinson committed:
Add logic around showing non-repeating category cards

kylehickinson committed:
Add additional buttons to welcome card

kylehickinson committed:
Allow group feed title labels (category, pub name, etc.) to wrap

kylehickinson committed:
Fix bug where Top News category wasn't at the top of the categories list

kylehickinson committed:
Ensure the first headline card is always Top News category item

kylehickinson committed:
Add check for empty parsed results for Brave Today resources

If the parsed result of a sources/feed JSON call end up with an empty array, it should not be counted as a successful call. It will now return an error in these cases

kylehickinson committed:
Add reset sources to default.

- Move source enabled logic to data source and update FeedItem.Source property name
- Add data source method to toggle entire category (to add in UI later)

kylehickinson committed:
Switch to inset grouped style for Brave Today settings (iOS 13+)

kylehickinson committed:
Remove trailing separators on source lists

kylehickinson committed:
Add ability to toggle entire category's sources on/off

kylehickinson committed:
Adjust brand title labels color and allow it to wrap

Also adjusts feed item date label text color slightly

kylehickinson committed:
Add empty state card with link to settings.

Rework action handler to support non feed item actions better and hookup the refresh button on the error card

kylehickinson committed:
Adjust color of brave today header and add a small shadow to it

kylehickinson committed:
Adjust scoring to lower frequency of repeating publishers

kylehickinson committed:
Adjust variety score calculation from 50% to 20% decrease per source hit

kylehickinson committed:
Don't load feed content in the bg if Brave Today is disabled

kylehickinson committed:
Refresh on NTP becoming visible rather than on created

Only if you are not scrolled into the feed

kylehickinson committed:
Handle feed UI based on state changes directly from the data source

- Move Brave Today related NTP UI into an overlay view that contains all 3 (loader, header, new content button)

kylehickinson committed:
Move heavy Brave Today operations onto a background thread

This fixes any hangs in the UI while parsing JSON, scoring items or generating cards

kylehickinson committed:
Add floating content available button when feed/sources expire

Also adds in check to reload feed when app is foregrounded

kylehickinson committed:
Don't hide the new content button when scrolling down during reload

kylehickinson committed:
Hook up refresh button on error card to attempt feed refresh

kylehickinson committed:
Reload cards when user has no cards being displayed and enables a source

kylehickinson committed:
Adjust variety score calculation from 20% to 5% decrease per source hit

kylehickinson committed:
Fix build failure caused by Deferred library move

kylehickinson committed:
Add legacy relative date time formatting for date labels on iOS 12

kylehickinson committed:
Replace hard-coded strings with localized equivalents

kylehickinson committed:
File cleanup & reorganization

kylehickinson committed:
Alter score computation slightly (server now providing base score value)

kylehickinson committed:
Remove unused file

iccub reviewed (changes requested):
Code looks nice overall, I still have to do high-level and functional overview

Please check why NTP VC doesn't deinitialize, must be retain cycle somewhere

And other than that, I noticed bunch runtime constraint errors, we can try to solve them if possible:

Few places where I noticed them:
- on opening new tab page
- on device orientation
- long press on an article
- tapping 'disable source' on an article

kylehickinson committed:
Restore cached sources on launch to ensure Today settings page is ready

kylehickinson committed:
Feedback part 1

kylehickinson committed:
Fix NTP ref cycle leak

Also removes static definition on the feed data source in BVC

kylehickinson committed:
Add support for deals categories

- Adds CategoryFillStrategy and uses it for both category group and deals
- Adds log for failed `FailableDecoable` decodes

kylehickinson committed:
Add description label to deals cards

kylehickinson committed:
Display the URL in context menus rather than the title

kylehickinson committed:
Add more brave offers button to deal cards

kylehickinson committed:
Fix Auto-Layout warnings logs when rotating orientation

kylehickinson committed:
Add localization clarification/comment for More Brave Offers label

iccub reviewed (changes requested):
few more comments regarding NTP VC

kylehickinson committed:
Feedback part 2

kylehickinson committed:
Fix scroll jumpiness when scrolling up after rotation

iccub reviewed (changes requested):
db review

kylehickinson committed:
Add fill strategy unit tests

kylehickinson committed:
Rename BraveTodaySourceMO to FeedSourceOverride

kylehickinson committed:
Fix not showing disable sources on iOS 12 legacy menus

kylehickinson committed:
Adds better accessibility support for Brave Today UI

iccub reviewed (approved):
Big PR, I did the best I could to review it, looks good, changing cards order and adding new card types is easy to do.

We are most likely going to wait with releasing it, I think we should merge it once ready and add a feature flag to hide it for now.

great job as always :)

kylehickinson committed:
Merge development into brave-today-feed

kylehickinson committed:
Merge branch 'development' into brave-today-feed

kylehickinson committed:
Merge branch 'development' into brave-today-feed

kylehickinson committed:
Add custom buckets for today resources

kylehickinson committed:
Merge branch 'development' into brave-today-feed
